### PR TITLE
Added obfuscation of yaml config, closes getlantern/lantern#3940

### PR DESCRIPTION
**Pull request by oxtoacart**

### CODE REVIEW

File: src/github.com/getlantern/yamlconf/yamlconf.go [hunk 1]
```diff
@@ -103,6 +103,14 @@ type Manager struct {
 	// example for fetching config updates from a remote server.
 	CustomPoll func(currentCfg Config) (mutate func(cfg Config) error, waitTime time.Duration, err error)
 
+	// Obfuscate: if true, the on-disk version of the config will be obfuscated by
+	// encrypting it using AES-128 encryption.
```
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Why this choice? Since it won't provide any kind of security, rot13+zip could've been sufficient. Not that important, but I wonder if a more lightweight solution could be possible.

File: src/github.com/getlantern/yamlconf/yamlconf.go [hunk 2]
```diff
@@ -148,25 +156,25 @@ func (m *Manager) Init() (Config, error) {
...
-		if err == nil {
-			_, err = m.saveToDiskAndUpdate(copied)
-		}
-		if err != nil {
-			return nil, fmt.Errorf("Unable to perform initial update of config on disk: %s", err)
+	// Always save whatever we loaded, which will cause defaults to be
+	// applied and formatting to be made consistent
+	copied, err := m.copy(m.cfg)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to copy config: %v", err)
```
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: We are using %v and %s inconsistently across this function for errors.

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 1 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Easy - golang has a built-in library for doing AES-128.  There's no built-in rot13 library.  Reading/writing AES-128 turns out to be 6 lines of code each.  I can live with that :)

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 2 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Fixed.

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 1 above)
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: I mean lightweight computationally. We are adding up miliseconds to startup by each new feature that we add, and that's how heavy programs get born ;)

File: src/github.com/getlantern/yamlconf/yamlconf_disk.go [hunk 3]
```diff
@@ -25,13 +31,30 @@ func (m *Manager) reloadFromDisk() (bool, error) {
...
+	var cfg Config
+	if m.Obfuscate {
+		log.Trace("Attempting to read obfuscated config")
+		var err1, err2 error
+		cfg, err1 = m.doReadFromDisk(true)
+		if err1 != nil {
+			log.Tracef("Error reading obfuscated config from disk, try reading non-obfuscated: %v", err)
+			cfg, err2 = m.doReadFromDisk(false)
+			if err2 != nil {
+				return false, err1
```
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf_disk.go]: In this case, why are we more interested in the first error? Perhaps return both concatenated?

File: src/github.com/getlantern/yamlconf/yamlconf_disk.go [hunk 4]
```diff
@@ -55,6 +78,37 @@ func (m *Manager) reloadFromDisk() (bool, error) {
...
 
+func (m *Manager) doReadFromDisk(allowObfuscation bool) (Config, error) {
+	infile, err := os.Open(m.FilePath)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to open config file %v for reading: %v", m.FilePath, err)
+	}
+	defer infile.Close()
+
+	var in io.Reader = infile
+	if allowObfuscation && m.Obfuscate {
```
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf_disk.go]: Why do you check m.Obfuscate here? If this function is called by `reloadFromDisk` and m.Obfuscate isn't true it won't read the obfuscated version in case it fails to read as non-obfuscated one.

Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf_disk.go]: I mean, it won't make line 53 very useful. Is that correct?

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 1 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Wow, this actually turns out to make a big difference.  Without encryption we're looking at about 300ms to read the config, with encryption about 1.6 seconds!  I'll do a custom ROT13 thing.

File: src/github.com/getlantern/yamlconf/yamlconf_disk.go (hunk 3 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf_disk.go]: Good idea, will fix.

File: src/github.com/getlantern/yamlconf/yamlconf_test.go [hunk 5]
```diff
@@ -184,17 +185,26 @@ func TestCustomPoll(t *testing.T) {
...
 }
 
-func assertSavedConfigEquals(t *testing.T, file *os.File, expected *TestCfg) {
+func assertSavedConfigEquals(t *testing.T, m *Manager, file *os.File, expected *TestCfg) {
 	b, err := yaml.Marshal(expected)
 	if err != nil {
 		t.Fatalf("Unable to marshal expected to yaml: %s", err)
 	}
-	bod, err := ioutil.ReadFile(file.Name())
+	stream, err := m.obfuscationStream()
```
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf_test.go]: With this, we are now only testing the obfuscated version path. Perhaps adding both would improve coverage. What do you think?

File: src/github.com/getlantern/yamlconf/yamlconf_disk.go (hunk 4 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf_disk.go]: Good point.

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 1 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Turns out those timings were a fluke.  I now see similar times for both (as low as 300ms and as high as 1.7 seconds).  So, I'm sticking with AES.

**Comment by oxtoacart** [comment]
@uaalto This is ready for another review.

File: src/github.com/getlantern/flashlight/config/config.go [hunk 6]
```diff
@@ -145,18 +118,14 @@ func Init(userConfig UserConfig, version string, configDir string, stickyConfig
 		log.Errorf("Could not get config path? %v", err)
 		return nil, err
 	}
-	run := isGoodConfig(configPath)
```
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/config.go]: @uaalto Until today, I hadn't realized that we were breaking encapsulation and reading/writing the config from within flashlight without using yamlconf.  That scheme obviously broke once we introduced obfuscation.  So, rather than apply obfuscation in both places, I cleaned this up so that everything is handled in yamlconf.

File: src/github.com/getlantern/yamlconf/yamlconf.go [hunk 7]
```diff
@@ -90,11 +90,20 @@ type Manager struct {
 	// FilePath: required, path to the config file on disk
 	FilePath string
 
-	// EmptyConfig: required, factor for new empty Configs
+	// ValidateConfig: optional. If specified, the config loaded from disk is
+	// validated with this function. If validation fails, the config is replaced
+	// with the result of DefaultConfig.
+	ValidateConfig func(cfg Config) error
```
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: New function that allows validating the config (needed by flashlight).

File: src/github.com/getlantern/yamlconf/yamlconf.go [hunk 8]
```diff
@@ -90,11 +90,20 @@ type Manager struct {
...
+
+	// DefaultConfig: optional. If specified, and there's a problem reading the
+	// config from disk or that , this is used.
+	DefaultConfig func() (Config, error)
```
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: New function that allows application of a default config (in our case, the embedded bootstrap config).

File: src/github.com/getlantern/flashlight/config/bootstrap.go [hunk 9]
```diff
@@ -102,14 +103,15 @@ func MakeInitialConfig(configPath string) error {
 	bytes, err := fs.GetIgnoreLocalEmpty("lantern.yaml")
```
Comment (uaalto) [inline src/github.com/getlantern/flashlight/config/bootstrap.go]: Are we ever reading the local filesystem for this? Why? Shouldn't the embedded be the default?

File: src/github.com/getlantern/flashlight/config/config.go [hunk 10]
```diff
@@ -77,35 +75,13 @@ type CA struct {
...
-	if fi, err := os.Stat(file); os.IsNotExist(err) {
-		log.Debugf("File does not exist at %v", file)
-		return fi, false
-	} else {
-		log.Debugf("File exists at %v", file)
-		return fi, true
+// hasCustomChainedServer returns an error if the config doesn't have a custom
+// chained server.
+func hasCustomChainedServer(_cfg yamlconf.Config) error {
+	cfg, ok := _cfg.(*Config)
```
Comment (uaalto) [inline src/github.com/getlantern/flashlight/config/config.go]: We are assuming here that just casting to a Config will provide a chained server. The case where we have a valid cast without a valid chained server is not going to happen?

File: src/github.com/getlantern/flashlight/config/config.go [hunk 11]
```diff
@@ -115,13 +91,10 @@ func hasCustomChainedServer(configPath, name string) bool {
...
 	// if it has been given a custom config with a custom chained server
 	// list
-	return nc > 0 && nc < 10
-}
-
-func isGoodConfig(configPath string) bool {
-	log.Debugf("Checking config path: %v", configPath)
-	fi, exists := exists(configPath)
-	return exists && hasCustomChainedServer(configPath, fi.Name())
+	if nc <= 0 || nc > 10 {
```
Comment (uaalto) [inline src/github.com/getlantern/flashlight/config/config.go]: ok I see. This is where probably my previous question gets answered. I would say that the function naming is a bit confusing, since what we are really checking there is if the config is valid, not if it has chained servers.

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 1 above)
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: I'm interested in how you did these measurements. Are they comparative differentials between versions of Lantern with and without this or a profiling of the function itself?
Normally symmetric encryption has some overhead, that's why I'm interested in this aspect.

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 2 above)
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Thanks.

File: src/github.com/getlantern/flashlight/config/bootstrap.go (hunk 9 above)
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/bootstrap.go]: This is using tarfs, which can include an embedded version.  This is existing code which I didn't change, so I'm assuming it does what it's supposed to.

File: src/github.com/getlantern/yamlconf/yamlconf_disk.go [hunk 12]
```diff
@@ -55,6 +79,69 @@ func (m *Manager) reloadFromDisk() (bool, error) {
...
+		}
+	}
+	return cfg, nil
+}
+
+func doReadFromDisk(filePath string, allowObfuscation bool, emptyConfig func() Config) (Config, error) {
+	start := time.Now()
+	infile, err := os.Open(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to open config file %v for reading: %v", filePath, err)
```
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf_disk.go]: %v / %s

File: src/github.com/getlantern/flashlight/config/config.go (hunk 11 above)
Comment (oxtoacart) [inline src/github.com/getlantern/flashlight/config/config.go]: Actually the only thing we've validating is that is has custom chained servers.  Still, at some point it may expand, so I've renamed it.

File: src/github.com/getlantern/flashlight/config/config.go (hunk 6 above)
Comment (uaalto) [inline src/github.com/getlantern/flashlight/config/config.go]: I agree it's a better abstraction.

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 1 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: I added some timing to the loadFromDisk method and printed out the results.  I just wrote a benchmark for thoroughness.  The benchmark reads a realistic config file from disk and decodes the yaml.  Depending on the run, the encryption adds anywhere between 0 and 13% overhead.  Without the yaml decoding the differences are much more severe, but the yaml decoding dwarfs the other work.

Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Oh, here's the benchmark - https://gist.github.com/oxtoacart/7e9b43efd08f6625a188bf48f8a2e0f3

File: src/github.com/getlantern/yamlconf/yamlconf_disk.go (hunk 12 above)
Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf_disk.go]: Left over from existing code, but I'll fix.

File: src/github.com/getlantern/yamlconf/yamlconf.go (hunk 1 above)
Comment (uaalto) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Impressive that you created the full benchmark. Thanks!!

Comment (oxtoacart) [inline src/github.com/getlantern/yamlconf/yamlconf.go]: Looks like rot13 by itself (no zip) is less overhead, only about 1.5%, and much more consistent.  I'll switch to that.

**Comment by oxtoacart** [comment]
Okay, I've switched to rot13 and merged from devel.  If tests all pass, I think this is good?

File: src/github.com/getlantern/rot13/rot13_test.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+func TestRoundTrip(t *testing.T) {
+	var in, out, rt bytes.Buffer
+	for i := byte(0); i <= 255; i++ {
+		err := in.WriteByte(i)
+		if !assert.NoError(t, err, "Unable to write byte") {
+			return
+		}
+		if i == 255 {
```
Comment (uaalto) [inline src/github.com/getlantern/rot13/rot13_test.go]: I'm curious, why did you decide to break at 255 instead of constraining the loop to `i < 255`?

Comment (oxtoacart) [inline src/github.com/getlantern/rot13/rot13_test.go]: Because I wanted to test all 256 possible values of a byte.  If I don't break at 255, i++ brings wraps back to 0, which meets the boundary condition, so this becomes an infinite loop.

Comment (uaalto) [inline src/github.com/getlantern/rot13/rot13_test.go]: Ah! I see, byte(0) actually creates a C char-like type. Of course, it makes sense.

**Comment by uaalto** [comment]
LGTM! Kudos for the benchmarking and all the tests in the code!!

**Comment by oxtoacart** [comment]
@uaalto Thanks for the review!
